File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/Chat/PanoChatInputView.m
//
//  PanoChatInputView.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoChatInputView.h"
#import "PanoGrowingTextView.h"
#import "PanoClientService.h"
#import "PanoDefine.h"

@implementation PanoChatInputView {
    
    UIView *inputContentView;
    
    UIStackView *controlView;
    
    UIButton *loudspeakerButton;
    
    UIButton *voiceButton;
    
    UIButton *earButton;
    
    UIButton *sendButton;
    
    UIStackView *sendStackView;
    
    void(^exitBlock)(void);
}

- (void)initViews {
    
    inputContentView = [UIView new];
    [self addSubview:inputContentView];
    
    textView = [[PanoGrowingTextView alloc] init];
    textView.placeholder = NSLocalizedString(@"聊两句", nil);
    textView.placeholderColor = [UIColor colorWithWhite:0.8 alpha:1.0];
    textView.minHeight = 35.0;
    textView.maxHeight = 35.0;
    textView.font = [UIFont systemFontOfSize:14];
    textView.backgroundColor = [UIColor whiteColor];
    textView.layer.cornerRadius = 4.0;
    [inputContentView addSubview:textView];
    
    controlView = [UIStackView new];
    controlView.axis = UILayoutConstraintAxisHorizontal;
    controlView.distribution = UIStackViewDistributionFillEqually;
    controlView.spacing = 25;
    [self addSubview:controlView];
    
    loudspeakerButton = [self createMusicButton:nil image:[UIImage imageNamed:@"room_loudspeaker"]];
    loudspeakerButton.selected = false;
    [loudspeakerButton setImage:[UIImage imageNamed:@"room_loudspeaker_close"] forState:UIControlStateSelected];
    loudspeakerButton.enabled = PanoIsPhone();
    
    voiceButton = [self createMusicButton:nil image:[UIImage imageNamed:@"room_voice"]];
    [voiceButton setImage:[UIImage imageNamed:@"room_voice_mute"] forState:UIControlStateSelected];
    
    earButton = [self createMusicButton:nil image:[UIImage imageNamed:@"room_ear_monitor"]];
    
    [loudspeakerButton addTarget:self action:@selector(startLoudspeaker) forControlEvents:UIControlEventTouchUpInside];
    [voiceButton addTarget:self action:@selector(enableAudio) forControlEvents:UIControlEventTouchUpInside];
    [earButton addTarget:self action:@selector(showEarMonitorPage) forControlEvents:UIControlEventTouchUpInside];
    
    [controlView addArrangedSubview:loudspeakerButton];
    
    sendStackView = [[UIStackView alloc] init];
    sendStackView.axis = UILayoutConstraintAxisHorizontal;
    sendButton = [self createMusicButton:NSLocalizedString(@" 发送 ", nil) image:nil];
    [sendButton setBackgroundImage:[UIImage imageNamed:@"home_bg_blue"] forState:UIControlStateNormal];
    [sendButton setBackgroundImage:[UIImage imageNamed:@"home_bg_blue"] forState:UIControlStateHighlighted];
    [sendButton addTarget:self action:@selector(sendText) forControlEvents:UIControlEventTouchUpInside];
    sendButton.layer.cornerRadius = 15;
    sendButton.layer.masksToBounds = true;
    sendButton.frame = CGRectMake(0, 2.5, 120, 30);
    [sendStackView addArrangedSubview:sendButton];
    sendStackView.hidden = true;
    [self addSubview:sendStackView];
    [self addNotifications];
}

- (UIButton *)createMusicButton:(NSString *)title image:(UIImage *)image {
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    if (title) {
        [button setTitle:title forState:UIControlStateNormal];
    }
    button.frame = CGRectMake(0, 0, 30, 30);
    if (image) {
        [button setImage:image forState:UIControlStateNormal];
        [button setImage:image forState:UIControlStateHighlighted];
    }
    return button;
}


- (void)initConstraints {
    [inputContentView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.left.mas_equalTo(self);
        make.height.mas_equalTo(60);
    }];
    
    [controlView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.mas_equalTo(controlView.superview).offset(-defaultMargin);
        make.height.mas_equalTo(40);
        make.left.mas_equalTo(inputContentView.mas_right).offset(8);
        make.centerY.mas_equalTo(textView);
    }];
    
    [sendStackView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.mas_equalTo(sendStackView.superview).offset(-defaultMargin);
        make.height.mas_equalTo(30);
        make.left.mas_equalTo(inputContentView.mas_right).offset(8);
        make.centerY.mas_equalTo(textView);
    }];
    
    [textView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(textView.superview).insets(UIEdgeInsetsMake(0, defaultMargin, 25, 8));
    }];
}

- (void)updateControlView:(BOOL)isHost chat:(BOOL)isChating {
    if (isHost) {
        if (![controlView.arrangedSubviews containsObject:voiceButton]) {
            [controlView addArrangedSubview:voiceButton];
        }
        if (![controlView.arrangedSubviews containsObject:earButton]) {
            [controlView addArrangedSubview:earButton];
        }
    } else {
        if (isChating) {
            if (![controlView.arrangedSubviews containsObject:voiceButton]) {
                [controlView addArrangedSubview:voiceButton];
            }
            if (![controlView.arrangedSubviews containsObject:earButton]) {
                [controlView addArrangedSubview:earButton];
            }
        } else {
            [voiceButton removeFromSuperview];
            [earButton removeFromSuperview];
        }
    }
    voiceButton.selected = PanoClientService.service.userService.me.audioStatus == PanoUserAudio_Mute;
    
}

- (void)sendText {
    if ([textView.text isEqualToString:@""]) {
        return;
    }
    [_delegate didSendText:textView.text];
    textView.text = @"";
    [textView resignFirstResponder];
}

- (void)startLoudspeaker {
    loudspeakerButton.selected = !loudspeakerButton.selected;
    [PanoClientService.service.audioService setLoudspeakerStatus:!loudspeakerButton.selected];
}

- (void)enableAudio {
    voiceButton.selected = [_delegate toggleAudio];
}

- (void)showEarMonitorPage {
    [_delegate showEarMonitoringView];
}

- (void)exitChat {
    [_delegate exitChat];
}

- (void)dealloc {
    [self removeNotifications];
}

#pragma mark --
- (void)addNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
}

- (void)removeNotifications {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardWillHideNotification object:nil];
}

- (void)keyboardWillShow:(NSNotification *)notification {
    CGSize keyboardSize = [[[notification userInfo] objectForKey:UIKeyboardFrameBeginUserInfoKey] CGRectValue].size;
    CGFloat keyboardHeight = keyboardSize.height;
    if ([_delegate respondsToSelector:@selector(keyboardWillShow:)]) {
        [_delegate keyboardWillShow:keyboardHeight];
    }
    [UIView animateWithDuration:0.3 animations:^{
        self->sendStackView.hidden = false;
        self->controlView.hidden = true;
    }];
}

- (void)keyboardWillHide:(NSNotification *)notification {
    [_delegate keyboardWillHide];
    [UIView animateWithDuration:0.3 animations:^{
        self->sendStackView.hidden = true;
        self->controlView.hidden = false;
    }];
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/Chat/PanoChatInputView.swift
//
//  PanoChatInputView.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/11/4.
//

import UIKit
import GrowingTextView

protocol PanoChatInputViewDelegate : NSObjectProtocol {
    
    func didSendText(text: String)
    
    func keyboardWillShow(height: CGFloat)
    
    func keyboardWillHide()
    
    func showEarMonitoringView();
    
    func exitChat();
    
    func toggleAudio() -> Bool;
}

class PanoChatInputView: PanoBaseView {
    
    var textView: GrowingTextView!
    
    var inputContentView: UIView!
    
    var controlView: UIStackView!
    
    var loudspeakerButton: UIButton!
    
    var voiceButton: UIButton!
    
    var earButton: UIButton!
    
//    var exitButton: UIButton!
    
    var sendButton: UIButton!
    
    var sendStackView: UIStackView!
    
    var exit: (() -> Void)? = nil
    
    weak var delegate: PanoChatInputViewDelegate?
    
    override func initViews() {
        
        inputContentView = UIView()
        self.addSubview(inputContentView)
        
        textView = GrowingTextView()
        textView.maxLength = 140
        textView.trimWhiteSpaceWhenEndEditing = false
        textView.placeholder = "聊两句".localized()
        textView.placeholderColor = UIColor(white: 0.8, alpha: 1.0)
        textView.minHeight = 35.0
        textView.maxHeight = 35.0
        textView.font = UIFont.systemFont(ofSize: 14)
        textView.backgroundColor = UIColor.white
        textView.layer.cornerRadius = 4.0
        inputContentView.addSubview(textView)
        
        controlView = UIStackView()
        controlView.axis = .horizontal
        controlView.distribution = .fillEqually
        controlView.spacing = 25;
        self.addSubview(controlView)
        
        loudspeakerButton = createMusicButton(image: UIImage(named: "room_loudspeaker"))
        loudspeakerButton.isSelected = false
        loudspeakerButton.setImage(UIImage(named: "room_loudspeaker_close"), for: .selected)
        loudspeakerButton.isEnabled = PanoIsPhone();
        voiceButton = createMusicButton(image: UIImage(named: "room_voice"))
        voiceButton.setImage(UIImage(named: "room_voice_mute"), for: .selected)
        earButton = createMusicButton(image: UIImage(named: "room_ear_monitor"))
//        exitButton = createMusicButton(image: UIImage(named: "room_exit_chat"))
        loudspeakerButton.addTarget(self, action: #selector(startLoudspeaker), for: .touchUpInside)
        voiceButton.addTarget(self, action: #selector(enableAudio), for: .touchUpInside)
        earButton.addTarget(self, action: #selector(showEarMonitorPage), for: .touchUpInside)
//        exitButton.addTarget(self, action: #selector(exitChat), for: .touchUpInside)
        controlView.addArrangedSubview(loudspeakerButton)
        
        sendStackView = UIStackView()
        sendStackView.axis = .horizontal
        sendButton = createMusicButton(title: "  发送  ".localized(), image: nil)
        sendButton.setBackgroundImage(UIImage(named: "home_bg_blue"), for: .normal)
        sendButton.setBackgroundImage(UIImage(named: "home_bg_blue"), for: .highlighted)
        sendButton.addTarget(self, action: #selector(sendText), for: .touchUpInside)
        sendButton.layer.cornerRadius = 15
        sendButton.layer.masksToBounds = true
        sendButton.frame = CGRect(x: 0, y: 2.5, width: 120, height: 30)
        sendStackView.addArrangedSubview(sendButton)
        sendStackView.isHidden = true
        self.addSubview(sendStackView)
        self.addNotifications()
    }
    
    func createMusicButton(title: String? = nil ,image: UIImage?) -> UIButton {
        let button = UIButton(type: .custom)
        if let title = title {
            button.setTitle(title, for: .normal);
        }
        button.frame = CGRect(x: 0, y: 0, width: 30, height: 30)
        button.setImage(image, for: .normal);
        button.setImage(image, for: .highlighted)
        return button
    }
    
    
    override func initConstraints() {
        inputContentView.snp.makeConstraints { (make) in
            make.bottom.left.equalToSuperview()
            make.height.equalTo(60)
        }
        controlView.snp.makeConstraints { (make) in
            make.right.equalToSuperview().offset(-defaultMargin)
            make.height.equalTo(40)
            make.left.equalTo(inputContentView.snp.right).offset(8)
            make.centerY.equalTo(textView)
        }
        
        sendStackView.snp.makeConstraints { (make) in
            make.right.equalToSuperview().offset(-defaultMargin)
            make.height.equalTo(30)
            make.left.equalTo(inputContentView.snp.right).offset(8)
            make.centerY.equalTo(textView)
        }
        
        textView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview().inset(UIEdgeInsets(top: 0, left: defaultMargin, bottom: 25, right: 8))
        }
    }
    
    func updateControlView(isHost: Bool, isChating: Bool) {
        if isHost {
            if !controlView.arrangedSubviews.contains(voiceButton) {
                controlView.addArrangedSubview(voiceButton)
            }
            if !controlView.arrangedSubviews.contains(earButton) {
                controlView.addArrangedSubview(earButton)
            }
            
        } else {
            if isChating {
                if !controlView.arrangedSubviews.contains(voiceButton) {
                    controlView.addArrangedSubview(voiceButton)
                }
                if !controlView.arrangedSubviews.contains(earButton) {
                    controlView.addArrangedSubview(earButton)
                }
//                if !controlView.arrangedSubviews.contains(exitButton) {
//                    controlView.addArrangedSubview(exitButton)
//                }
            } else {
                voiceButton.removeFromSuperview()
                earButton.removeFromSuperview()
//                exitButton.removeFromSuperview()
//                print("controlView.arrangedSubviews-> \(controlView.arrangedSubviews)")
            }
        }
        
        voiceButton.isSelected = PanoUserService.service()?.me()?.audioStatus == .mute
        
    }
    @objc func sendText() {
        if textView.text.isEmpty {
            return
        }
        delegate?.didSendText(text: textView.text)
        textView.text = ""
        textView.resignFirstResponder()
    }
    
    @objc func startLoudspeaker() {
        loudspeakerButton.isSelected = !loudspeakerButton.isSelected
        PanoAudioService.service()?.setLoudspeakerStatus(staus: !loudspeakerButton.isSelected)
    }
    
    @objc func enableAudio() {
        voiceButton.isSelected = self.delegate!.toggleAudio()
//        voiceButton.isSelected = self.toggleAudioBlock!();
//        voiceButton.isSelected = !voiceButton.isSelected
//        if voiceButton.isSelected {
//            PanoAudioService.service()?.muteAudio()
//        } else {
//            PanoAudioService.service()?.unmuteAudio()
//        }
    }
    
    @objc func showEarMonitorPage() {
        delegate?.showEarMonitoringView()
    }
    
    @objc func exitChat() {
        delegate?.exitChat();
    }
    
    deinit {
        self.removeNotifications()
    }
}


extension PanoChatInputView {
    
    func addNotifications() {
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(notification:)), name: UIApplication.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide(notification:)), name: UIApplication.keyboardWillHideNotification, object: nil)
    }
    
    func removeNotifications() {
            NotificationCenter.default.removeObserver(self, name: UIApplication.keyboardWillShowNotification, object: nil)
            NotificationCenter.default.removeObserver(self, name: UIApplication.keyboardWillShowNotification, object: nil)
    }
    
    @objc func keyboardWillShow(notification: Notification) {
        let keyboardSize = (notification.userInfo![UIResponder.keyboardFrameEndUserInfoKey] as! NSValue).cgRectValue.size
        let keyboardHeight = keyboardSize.height;
        delegate?.keyboardWillShow(height: keyboardHeight)
        UIView.animate(withDuration: 0.3) {
            self.sendStackView.isHidden = false
            self.controlView.isHidden = true
        }
    }
    
    @objc func keyboardWillHide(notification: Notification) {
        delegate?.keyboardWillHide()
        UIView.animate(withDuration: 0.3) {
            self.sendStackView.isHidden = true
            self.controlView.isHidden = false
        }
    }
}
